pkg/apis/batch/validation/validation.go

```diff
@@ -565,6 +567,35 @@ func ValidateJobTemplateSpec(spec *batch.JobTemplateSpec, fldPath *field.Path, o
 	return allErrs
 }
 
+func validateCompletions(spec, oldSpec batch.JobSpec, fldPath *field.Path) field.ErrorList {
+	if !feature.DefaultFeatureGate.Enabled(features.ElasticIndexedJob) {
+		return apivalidation.ValidateImmutableField(spec.Completions, oldSpec.Completions, fldPath)
+	}
+	// Completions is immutable for non-indexed jobs.
+	// For Indexed Jobs, if ElasticIndexedJob feature gate is not enabled,
+	// fall back to validating that spec.Completions is always immutable.
+	var allErrs field.ErrorList
+	isIndexedJob := spec.CompletionMode != nil && *spec.CompletionMode == batch.IndexedCompletion
+	if !isIndexedJob {
+		return apivalidation.ValidateImmutableField(spec.Completions, oldSpec.Completions, fldPath)
+	}
+
+	// Indexed Jobs cannot set completions to nil. The nil check
+	// is already performed in validateJobSpec, no need to add another error.
+	if spec.Completions == nil || oldSpec.Completions == nil {
```

Comment:
I'm unsure about skipping the rest of this function if `oldSpec.Completions == nil`

I would do this instead:

```go
if apiequality.Semantic.DeepEqual(spec.Completions, oldSpec.Completions) {
  return allErrs
}

if spec.Completions == nil {
  allErrs = append(allErrs, apivalidation.ValidateImmutableField(spec.Completions, oldSpec.Completions, fldPath))
  return allErrs
}

if *spec.Completions != *spec.Parallelism {
  allErrs = append(allErrs, field.Invalid(fldPath, spec.Completions, fmt.Sprintf("can only be modified in 
  tandem with %s", fldPath.Root().Child("parallelism").String())))
}
```